### PR TITLE
[Attention] Enable blockwise transposes/rotations to avoid bank conflicts.

### PR DESCRIPTION
**Pull request by manupak**
This commit mainly adds support for thread<->iter tranposes and K<->D rotations
when storing/reading from LDS to avoid bank conflicts in attention kernels (similar to what we do for gemms)

In order to do the above, firstly, it refactors out the said utilities in the gemm lowering.
Then use them in attention lowering.

closes : https://github.com/ROCm/rocMLIR-internal/issues/1449

TODO:
* ~~post e2e perf~~
* post bank conflict data

### CODE REVIEW

**Comment by manupak** [comment]
<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:///C:/Users/mkarunar/AppData/Local/Temp/msohtmlclip1/01/clip.htm">
<link rel=File-List
href="file:///C:/Users/mkarunar/AppData/Local/Temp/msohtmlclip1/01/clip_filelist.xml">
<style>

</style>
</head>

<body link="#467886" vlink="#96607D">


  | #develop |   | #ldsrotate | %
-- | -- | -- | -- | --
-with-attn-scale false -transO false   -transV false -transK true -transQ false -t f16 -g 256 -seq_len_q 1024   -seq_len_k 1024 -head_dim_qk 128 -head_dim_v 128 | 87.22489 |   | 87.80448371 | 0.66448
-with-attn-scale false -transO false   -transV false -transK true -transQ false -t f16 -g 128 -seq_len_q 2048   -seq_len_k 2048 -head_dim_qk 128 -head_dim_v 128 | 93.98149 |   | 93.97638092 | -0.00544
-with-attn-scale false -transO false   -transV false -transK true -transQ false -t f16 -g 64 -seq_len_q 4096   -seq_len_k 4096 -head_dim_qk 128 -head_dim_v 128 | 98.04433 |   | 97.93254757 | -0.11401
-with-attn-scale false -transO false   -transV false -transK true -transQ false -t f16 -g 32 -seq_len_q 8192   -seq_len_k 8192 -head_dim_qk 128 -head_dim_v 128 | 99.52962 |   | 99.61194133 | 0.082715
-with-attn-scale false -transO false   -transV false -transK true -transQ false -t f16 -g 16 -seq_len_q 16384   -seq_len_k 16384 -head_dim_qk 128 -head_dim_v 128 | 101.4314 |   | 101.4381675 | 0.006689
-with-attn-scale false -transO false   -transV false -transK true -transQ false -t f16 -g 12 -seq_len_q 384   -seq_len_k 384 -head_dim_qk 64 -head_dim_v 64 | 13.4178 |   | 14.22691055 | 6.030151
-with-attn-scale false -transO false   -transV false -transK true -transQ false -t f16 -g 768 -seq_len_q 384   -seq_len_k 384 -head_dim_qk 64 -head_dim_v 64 | 67.50876 |   | 67.58430078 | 0.111898
-transO false -transV false -transK true   -transQ false -with-attn-scale false -g 40 -seq_len_q 1024 -seq_len_k 1024   -head_dim_qk 64 -head_dim_v 64 -t f16 | 62.36883 |   | 62.31092293 | -0.09285
-transO false -transV false -transK true   -transQ false -with-attn-scale false -g 40 -seq_len_q 1024 -seq_len_k 77   -head_dim_qk 64 -head_dim_v 64 -t f16 | 21.38251 |   | 21.56526496 | 0.854701
-transO false -transV false -transK true   -transQ false -with-attn-scale false -g 20 -seq_len_q 4096 -seq_len_k 4096   -head_dim_qk 64 -head_dim_v 64 -t f16 | 73.59418 |   | 73.66480109 | 0.095962
-transO false -transV false -transK true   -transQ false -with-attn-scale false -g 20 -seq_len_q 4096 -seq_len_k 77   -head_dim_qk 64 -head_dim_v 64 -t f16 | 26.69985 |   | 26.62940369 | -0.26385
-transO false -transV false -transK true   -transQ false -with-attn-scale false -t f16 -g 1 -seq_len_q 16384 -seq_len_k   16384 -head_dim_qk 512 -head_dim_v 512 | 54.34416 |   | 63.17845024 | 16.2562
-transO false -transV false -transK true   -transQ false -with-attn-scale false -t f32 -g 1 -seq_len_q 16384 -seq_len_k   16384 -head_dim_qk 512 -head_dim_v 512 | 21.58675 |   | 23.24375919 | 7.676046



</body>

</html>

File: mlir/lib/Dialect/Rock/Transforms/GridwiseGemmToBlockwise.cpp [hunk 1]
```diff
@@ -2182,7 +2249,7 @@ struct GridwiseAttentionAccelRewritePattern
               gemm0OutSubTileNxMViews, gemm0ExpOutBufferToLDS,
               gemm1LDSByteBufferB, gemm1KpacksPerBlock, "n", gemm1KPerBlock,
               gemm1NPerBlock, /*copyKPerThread=*/1, gemm1InNPerThread,
-              forceUnroll);
+              forceUnroll, false);
```
Comment (krzysz00) [inline mlir/lib/Dialect/Rock/Transforms/GridwiseGemmToBlockwise.cpp:2250]: What's this last parameter?

**Review by krzysz00 — APPROVED** [review]
If I'm reading the PR right, the refactored functions are being used up in GridwiseGemmACcell as well.

Approved

**Comment by krzysz00** [comment]
@manupak Do you want to land this? I think you want to land this

**Comment by manupak** [comment]
I d be comfortable not landing this in 6.2 -- because I think it has gains from other PR (perf config PR) in it (I d need to double check this) -- so that gives the last two boost larger head_dim case. Thus, it only leaves us with 6% bump in a single case which averages out to almost nothing.

Therefore, the risk it introduces seems larger for something that does not improve perf that much.

**Pull request merged**